### PR TITLE
Initial implementation of type McCabe (#683)

### LINKED ISSUE #683: Cyclomatic Complexity at Type Level for C++
See #682 for *Cyclomatic Complexity* at Function Level.

Adapted to the object-oriented paradigm, this metric can also be defined for classes as the sum of its methods complexity metric.

### PR DESCRIPTION
**Pull request by Seeker04**
TODO checklist:
- [x] Write tests - [3fdb63c](https://github.com/Ericsson/CodeCompass/pull/727/commits/3fdb63c4c06783e883f654f5fd285bc566239ec7)
- [x] Handling of nested classes
- [x] Wait for decision on handling of template classes - [0bf929a](https://github.com/Ericsson/CodeCompass/pull/727/commits/0bf929af35737f46615d9a8236ccbe5cf08092fd)
- [x] Check validity of non-zero values for classes with no user-defined methods - [c3cf55e](https://github.com/Ericsson/CodeCompass/pull/727/commits/c3cf55e664b60dd7435c33bf013cc040cf363137)

Closes #683

### CODE REVIEW

**Comment by mcserep** [comment]
**Handling of nested classes:** for now, we should handle nested classes as separate, independent types and not aggregate the complexity.

 **Wait for decision on handling of template classes:** investigate the reason for different values for the same types, e.g. `DynArray` mentioned at https://github.com/Ericsson/CodeCompass/issues/683#issuecomment-2047469351 .

File: plugins/cpp_metrics/model/include/model/cppastnodemetrics.h [hunk 1]
```diff
@@ -17,9 +17,10 @@ struct CppAstNodeMetrics
   {
     PARAMETER_COUNT = 1,
     MCCABE = 2,
-    BUMPY_ROAD = 3,
-    LACK_OF_COHESION = 4,
-    LACK_OF_COHESION_HS = 5,
+    MCCABE_TYPE = 3,
```
Comment (mcserep) [inline plugins/cpp_metrics/model/include/model/cppastnodemetrics.h:20]: In this case I suggest the `MCCABE` to be renamed as well.

File: plugins/cpp_metrics/test/sources/parser/CMakeLists.txt [hunk 2]
```diff
@@ -3,5 +3,6 @@ project(CppMetricsTestProject)
 
 add_library(CppMetricsTestProject STATIC
   mccabe.cpp
+  typemccabe.cpp
```
Comment (mcserep) [inline plugins/cpp_metrics/test/sources/parser/CMakeLists.txt:6]: Maybe also rename the `mccabe.cpp` as well.

**Review by mcserep — COMMENTED** [review]
Looks quite good based on the code. Requires further functional testing to complete the review.

File: plugins/cpp_metrics/parser/src/cppmetricsparser.cpp [hunk 3]
```diff
@@ -275,6 +363,7 @@ bool CppMetricsParser::parse()
   functionParameters();
   functionMcCabe();
   functionBumpyRoad();
+  typeMcCabe();
```
Comment (mcserep) [inline plugins/cpp_metrics/parser/src/cppmetricsparser.cpp:366]: Please add LOG statement.

File: plugins/cpp_metrics/model/include/model/cppastnodemetrics.h (hunk 1 above)
Comment (Seeker04) [reply plugins/cpp_metrics/model/include/model/cppastnodemetrics.h:20]: Solved in [62406a3](https://github.com/Ericsson/CodeCompass/pull/727/commits/62406a35283dc77fc5046950a96f49f252c9dbda).

File: plugins/cpp_metrics/test/sources/parser/CMakeLists.txt (hunk 2 above)
Comment (Seeker04) [reply plugins/cpp_metrics/test/sources/parser/CMakeLists.txt:6]: Solved in [62406a3](https://github.com/Ericsson/CodeCompass/pull/727/commits/62406a35283dc77fc5046950a96f49f252c9dbda).

File: plugins/cpp_metrics/parser/src/cppmetricsparser.cpp (hunk 3 above)
Comment (Seeker04) [reply plugins/cpp_metrics/parser/src/cppmetricsparser.cpp:368]: Rebased and added the LOG statement.

**Review by mcserep — APPROVED** [review]

**Review by intjftw — APPROVED** [review]

**Pull request merged**